### ISSUE TITLE
Add more features

### ISSUE THREAD

**Issue by boukeversteegh** [datastack-net/dockerized#7]
Currently there are almost no features, and they are implemented twice, once in Bash and once in Powershell.
This is important to avoid any dependencies beyond Docker, and to make it cross-platform.

In the long term, implementing everything twice in languages like Bash and Powershell may not be sustainable.

How should Dockerized be implemented, to support more features, keep it cross-platform, and make it easy to contribute and maintain?

Here are all the options I've thought of, happy to hear your thoughts.

- Platform supported scripts (Bash + Powershell)
  - ✔️ Natively supported by platform without dependencies
  - 👎 Requires maintaining 2 implementations
  - 👎 Difficult to maintain as scripts become more complex
  - 👎 Not accessible for contributors when scripts turn into full programs
  - 👎 Hard to test (no unit tests, debugging)
- Scripts: e.g. Python / Javascript
  - ✔️ Easy to use for developers.
  - ✔️ Great free IDE (PyCharm)
  - :x: Requires an interpreter to be installed. This is a deal-breaker, as dockerized should run without dependencies (other than docker+compose).
- Scripts packaged into executables:
  - ⁉ Is there any packager that can cross-compile?
    - :x: Github actions does not host all platforms (e.g. arm64), so this would mean `dockerized` cannot be used everywhere.
  - Python + PyInstaller
    - :x: Cannot cross-compile. You need the target platform to compile to it. 
  - ...?
- Compiled languages:
   - ✔️ Single executable without dependencies.
   - 👎 Requires managing releases and versions, compiling to multiple target platforms.
     - Semantic release bot and Github actions can automate all of this.
  - Go:
    - ✔️ Small binaries.
    - ✔️ Cross-platform compilation possible.
    - ✔️ Docker-compose is written in Go, potentially able to reuse libraries.
    - ✔️ For users who have go installed, can easily install `dockerized` with `go get`.
    - 👎 Limited IDE options (Goland is not free, VsCode limited compared to full IDE)
    - 👎 I've never written any application in Go, and have no clue about best practices.
    - 👎 Somewhat low-level language. Fewer potential collaborators compared to python or node.
  - C#:
    - ✔️ Easy to pick up OOP language
    - ✔️ Cross-platform compilation possible.
    - ✔️ Free IDE available (Visual Studio)
    - ✔️ My main language at the moment
    - 👎 Recommended IDE (Rider) not free
    - 👎 Not as mainstream as Python or Node.
    - 👎 Big binaries as they need to include the dotnet sdk (58mb)
    - 👎 Not very mainstream.
- Script language executed within docker (e.g. dockerized python `bin/dockerized.py`)
   - ✔️ Any language can be chosen. It will run in docker so all dependencies can be there.
   - 👎 The script will need to call docker-compose from within the container. I'm afraid this will cause problems.
   - :x: This means 2 calls to `docker start`. On my super fast Windows system `docker start` takes 1.5 seconds. Dockerized would become unbearably slow.
- Compiled Language (Go, C#), pre-compiled within docker, and executed on host  
  > ℹ️ E.g. Include `dockerized.go`, launch a `golang` container that compiles an .exe for Windows, extract it, and run it on the host. This can be done for all platforms
  - ✔️ Same benefits as all other compiled language options
  - ✔️ No releases or versioning needed
  - 👎 Additional logic for the compilation process in bash and powershell
  - ❔ Not sure if it's possible to compile 
- Any language, and auto-download a standalone interpreter onto the host
  > Let `dockerized` figure out the platform, download the executable into a local directory and then continue in the other language
  - :x: Python - Does not publish prebuilt binaries for unix platforms (only packages in repositories)
    - A project that attempts to package python https://python-build-standalone.readthedocs.io/en/latest/ warns that the Windows builds are 'extremely unstable'
  - ✔️ Node - Zips/Tars for all platforms and architectures, including windows. Small size (20-50mb)
    - `npm install` would need to run on the host, as in some cases precompiled binaries are included.
    - ☹️ 32 bit linux not supported anymore
  - ✔️ Go - Available for all platforms as Zip, but quite big (~100Mb)

**Comment by boukeversteegh** [comment]
I've managed to port the current feature-set to go.

Please try it out. Powershell not yet implemented:

```bash
git checkout feature/go
dockerized --help
```

The first time you run `dockerized`, it will compile, and then run the new go-implementation.

I've added some useful debugging features that I wanted to add before, but couldn't:

```bash
# rebuild the docker image before running. e.g. dockerize --build tree
dockerized --build <command>

# open a shell in the container for the command (opens bash if available, sh otherwise)
dockerized --shell node
```

To change the go implementation, you can just change the source files and start dockerized like this:

```bash
dockerized --compile [options] <command>
```

This will re-compile the go-code and run dockerized. It was necessary to implement this helper command, because `go run` would run the binary in a temp folder, breaking the relative location of `docker-compose.yml`.

**Comment by boukeversteegh** [comment]
Go port is now on `master`. Everything should work out of the box, but if anyone has trouble, please open an issue.

Merged in #8
